src/scenic/core/lazy_eval.py

```diff
@@ -13,174 +13,195 @@
-		import scenic.syntax.veneer as veneer
-		if veneer.simulationInProgress() and veneer.currentBehavior:
-			agent = veneer.currentBehavior._agent
-			assert agent
-			darg.__init__(*args, **kwargs)
-			agent._evaluated = DefaultIdentityDict()
-			value = darg.evaluateIn(agent)
-			del agent._evaluated
-			return value
-		else:
-			return darg
-
-	def __init__(self, requiredProps, value, _internal=False):
-		self.value = value
-		super().__init__(requiredProps)
-
-	def evaluateInner(self, context):
-		return self.value(context)
-
-	def __getattr__(self, name):
-		if name.startswith('__') and name.endswith('__'):	# ignore special attributes
-			return object.__getattribute__(self, name)
-		return DelayedArgument(self._requiredProperties,
-			lambda context: getattr(self.evaluateIn(context), name),
-			_internal=True)
-
-	def __call__(self, *args, **kwargs):
-		subprops = (requiredProperties(arg) for arg in itertools.chain(args, kwargs.values()))
-		props = self._requiredProperties.union(*subprops)
-		def value(context):
-			subvalues = (valueInContext(arg, context) for arg in args)
-			kwsvs = { name: valueInContext(arg, context) for name, arg in kwargs.items() }
-			return self.evaluateIn(context)(*subvalues, **kwsvs)
-		return DelayedArgument(props, value, _internal=True)
+    """DelayedArgument(requiredProps, value, _internal=False)
+
+    Specifier arguments requiring other properties to be evaluated first.
+
+    The value of a DelayedArgument is given by a function mapping the context (object under
+    construction) to a value.
+
+    .. note::
+
+        When called from a dynamic behavior, constructors for delayed arguments return
+        *actual evaluations*, not `DelayedArgument` objects. The agent running the
+        behavior is used as the evaluation context.
+
+    Args:
+        requiredProps: see `LazilyEvaluable`.
+        value: function taking a single argument (the context) and returning the
+            corresponding evaluation of this object.
+        _internal (bool): set to `True` for internal uses that need to suppress the
+            exceptional handling of calls from dynamic behaviors above.
+    """
+    def __new__(cls, *args, _internal=False, **kwargs):
+        darg = super().__new__(cls)
+        if _internal:
+            return darg
+        # at runtime, evaluate immediately in the context of the current agent
+        import scenic.syntax.veneer as veneer
+        if veneer.simulationInProgress() and veneer.currentBehavior:
+            agent = veneer.currentBehavior._agent
+            assert agent
+            darg.__init__(*args, **kwargs)
+            agent._evaluated = DefaultIdentityDict()
+            value = darg.evaluateIn(agent)
+            del agent._evaluated
+            return value
+        else:
+            return darg
+
+    def __init__(self, requiredProps, value, _internal=False):
+        self.value = value
+        self.suppressModifying = (len(inspect.signature(value).parameters) == 1)
```

Comment:
Per our discussion, remove this and the `modifying` argument (both here and in `valueInContext`, etc.) since we ended up not using it.